src/main/java/com/orange/fiware/openlpwa/ngsi/NgsiManager.java

```diff
@@ -0,0 +1,208 @@
+     * @throws AgentException when the subscriptionId is null
+     */
+    public ListenableFuture<UpdateContextSubscriptionResponse> updateSubscription(String subscriptionId) throws AgentException {
+        if (subscriptionId == null) {
+            String errorMsg = "Unable to update subscription for a null subscriptionId";
+            logger.error(errorMsg);
+            throw new AgentException(errorMsg);
+        }
+        UpdateContextSubscription updateContextSubscription = new UpdateContextSubscription();
+        updateContextSubscription.setSubscriptionId(subscriptionId);
+        updateContextSubscription.setDuration("P1M");
+        return ngsiClient.updateContextSubscription(contextBrokerRemoteUrl, remoteHeaders(), updateContextSubscription);
+    }
+
+    /**
+     * Unsubcribes a device
+     *
+     * @param subscriptionId Subscription identifier
+     * @return A future for a UnsubscribeContextResponse
+     * @throws AgentException when the subscriptionId is null
+     */
+    public ListenableFuture<UnsubscribeContextResponse> unsubscribe(String subscriptionId) throws AgentException {
+        if (subscriptionId == null) {
+            String errorMsg = "Unable to unsubscribe for a null subscriptionId";
+            logger.error(errorMsg);
+            throw new AgentException(errorMsg);
+        }
+        return ngsiClient.unsubscribeContext(contextBrokerRemoteUrl, remoteHeaders(), subscriptionId);
+    }
+
+    /**
+     * Updates device attributes sending an updateContext request to the context broker with the deviceID
+     *
+     * @param deviceID      Device identifier
+     * @param attributeList Attributes to update
+     * @return A future for UpdateContextResponse
+     * @throws AgentException when the deviceID is null or where there isn't an attribute to update
+     */
+    public ListenableFuture<UpdateContextResponse> updateDeviceAttributes(String deviceID, List<ContextAttribute> attributeList) throws AgentException {
+        UpdateContext context = new UpdateContext();
+        context.setUpdateAction(UpdateAction.APPEND);
+        List<ContextElement> elementList = new ArrayList<>();
+        ContextElement element = new ContextElement();
+        element.setEntityId(new EntityId(deviceID, "", false));
+        element.setContextAttributeList(attributeList);
+        elementList.add(element);
+        context.setContextElements(elementList);
+        logger.info("Sending an update request to the context broker (deviceID:{}, context:{}, list:{})", deviceID, context, attributeList);
+        return ngsiClient.updateContext(contextBrokerRemoteUrl, remoteHeaders(), context);
+    }
+
+    /**
+     * Generates headers for Context Broker
+     *
+     * @return Http headers for provided configuration
+     */
+    private HttpHeaders remoteHeaders() {
+        HttpHeaders httpHeaders = new HttpHeaders();
+
+
+        if (contextBrokerRemoteAuthToken == null) {
+            setAccessTokenSync();
+        }
```

Comment:
setAccessToken() may not find a token. In this case, the header must not contain an "X-Auth-Token". The state of contextBrokerRemoteAuthToken must therefore be re-evaluated: this explains the "if" rather than an "else".